design-decisions/2018-07-25 Attributes as patterns.md

```diff
@@ -0,0 +1,120 @@
+Attributes as Patterns
+======================
+
+Background
+----------
+
+Attributes are used in Kore in order to provide hints for manipulating symbols,
+axioms, sorts and so on.
+When translating from K, the Kore attributes are a
+fairly direct translation of the K ones.
+
+Each Kore backend can have its own set of attributes, and backends must be able
+to ignore unknown attributes.
+
+Examples include the `constructor` attribute which
+specifies that a certain symbol is a constructor, so, say, it
+cannot be unified with different constructors.
+
+Questions asked
+---------------
+
+1. What kind of structure do attributes have?
+1. What syntax should we use?
+1. How can we make sure that attributes have some documentation?
+
+The answer to the first two questions comes from a Slack discussion on
+`runtimeverification/#haskell`(https://runtimeverification.slack.com/messages/C9B9QBF46/) on 2018-17-25
+
+* between @grosu, @bmmoore, @dwightguth, @u--, @njohnwalker, @traiansf, @hreada.
+* Starting at: https://runtimeverification.slack.com/archives/C9B9QBF46/p1532543176000083
+* Ending at: https://runtimeverification.slack.com/archives/C9B9QBF46/p1532562323000104
+
+The answer to the third
+comes from a face-to-face discussion between @traiansf and @virgil-serbanuta.
+
+Attribute structure
+-------------------
+
+Attributes are identified by a name and might have (or not) parameters.
+
+When present, attribute parameters are usually very simple, e.g. the `strict` attribute
+can have a list of argument indexes in which it is strict.
+
+Sometimes we may
+want to refer to user-defined symbols, e.g. we may use the `unit` attribute
+to specify the neutral element for an operation, and we may want to use this
+attribute both for the operation symbol declaration and for the axioms that
```

Comment:
In K and ML, symbols are not necessarily "operation" symbols.  In FOL, "operation symbols" are interpreted as total functions in models, while "predicate symbols" as relations.  So stick to just "symbols".

Reply:
Solved.